src/libraries/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Certificate2.cs

```diff
@@ -81,6 +92,8 @@ public X509Certificate2(byte[] rawData, SecureString? password, X509KeyStorageFl
         ///   The certificate data to process.
         /// </param>
         /// <exception cref="CryptographicException">An error with the certificate occurs.</exception>
+        [UnsupportedOSPlatform("ios")]
+        [UnsupportedOSPlatform("tvos")]
         public X509Certificate2(ReadOnlySpan<byte> rawData)
```

Comment:
[SecCertificateCreateWithData](https://developer.apple.com/documentation/security/1396073-seccertificatecreatewithdata?language=objc) says it works on iOS 2.0+, I don't think most of these ctors should be marked as unsupported.  (Worst case, some of the file formats don't work, but the API will still work in most cases)

Reply:
Thanks, same as above: just fixed the build failures due to the chain of invocations coming from unsupported APIs. Perhaps, it needs more fine tuning.

Reply:
I've tried to remove some redundant annotations

Reply:
@bartonjs The calls into `GetKeys` methods [here](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ApplePkcs12Reader.cs#L78-L96) is the cause of build failures now. What is a better approach to fix it - either to wrap problematic calls into `if (OperatingSystem.IsIOS() || OperatingSystem.IsTvOS()) { // what to return here? }` or  in those `GetKeys` methods?

Reply:
> What is a better approach to fix it

That depends -- what is the behavior supposed to be?

If the plan is ship the PNSE (or we're just tracking work with this attribute) then paint the failing method... there's not really an implementation of GetRSAPrivateKey() that can work without being able to get a key handle of some kind.

If this change is supposed to produce the definitive set of support attributes then... well, honestly, if it's supposed to be definitive then I imagine that a lot of it needs to be thrown out.  e.g. SecKeyGeneratePair is supported as of iOS 2.0, so I don't see why AppleCryptoNative_RsaGenerateKey can't work (maybe it happens to currently use some implementation details that need to vary), which means that most of RSASecurityTransforms would no longer be painted as unsupported, etc.

Reply:
I assume it's rather the first option - just to reflect the fact that the part of APIs is excluded in the native code for iOS/tvOS. I want to reach this by excluding unsupported APIs from the library build and putting PNSE where it could be required. In turn, it might need respective annotations with `UnsupportedOSPlatform` attribute w/o any details of OS version, b/c, I guess, the native part is disabled for iOS/tvOS in general.

Reply:
> b/c, I guess, the native part is disabled for iOS/tvOS in general.

In the case of the RSA operations, I think it's because the shim was made to compile by using a chainsaw instead of a scalpel... anything with an error was just \#if'd out.  I assume a whole lot of stuff that currently doesn't work would work just fine, but perhaps requiring some tactical \#ifs to deal with macOS/Keychain vs iOS differences.

Reply:
As an example, I tried to enable [AppleCryptoNative_RsaGenerateKey](https://github.com/dotnet/runtime/blob/main/src/libraries/Native/Unix/System.Security.Cryptography.Native.Apple/pal_rsa.c#L10), but it just led to such build errors as:
```
...
/Users/maxim/dev/runtime/src/libraries/Native/Unix/System.Security.Cryptography.Native.Apple/pal_rsa.c(13,26): error G812D74F5: 'SecKeychainRef' is unavailable: not available on iOS [/Users/maxim/dev/runtime/src/libraries/Native/build-native.proj]
/Users/maxim/dev/runtime/src/libraries/Native/Unix/System.Security.Cryptography.Native.Apple/pal_rsa.c(34,42): error GED893B77: 'kSecUseKeychain' is unavailable: not available on iOS [/Users/maxim/dev/runtime/src/libraries/Native/build-native.proj]
...
``` 
`SecKeychainRef` refers to `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Security.framework/Versions/A/Headers/SecBase.h` 
```
typedef struct CF_BRIDGED_TYPE(id) __SecKeychain *SecKeychainRef
    API_AVAILABLE(macos(10.0)) API_UNAVAILABLE(ios, tvos, watchos);
```

Reply:
Yeah, that's the chainsaw/scalpel.

Something like the following feels like it would work

```C
#if defined(TARGET_MACCATALYST) || defined(TARGET_IOS) || defined(TARGET_TVOS)
#define NO_KEYCHAIN 1
#endif

int32_t AppleCryptoNative_RsaGenerateKey(
    int32_t keySizeBits,
#ifdef NO_KEYCHAIN
    void* ignored,
#else
    SecKeychainRef tempKeychain,
#endif
    SecKeyRef* pPublicKey, SecKeyRef* pPrivateKey, int32_t* pOSStatus)
{
    if (pPublicKey != NULL)
        *pPublicKey = NULL;
    if (pPrivateKey != NULL)
        *pPrivateKey = NULL;

    if (pPublicKey == NULL || pPrivateKey == NULL || pOSStatus == NULL)
        return kErrorBadInput;
    if (keySizeBits < 384 || keySizeBits > 16384)
        return -2;

    CFMutableDictionaryRef attributes = CFDictionaryCreateMutable(NULL, 3, &kCFTypeDictionaryKeyCallBacks, NULL);

    CFNumberRef cfKeySizeValue = CFNumberCreate(NULL, kCFNumberIntType, &keySizeBits);
    OSStatus status;

    if (attributes != NULL && cfKeySizeValue != NULL)
    {
        CFDictionaryAddValue(attributes, kSecAttrKeyType, kSecAttrKeyTypeRSA);
        CFDictionaryAddValue(attributes, kSecAttrKeySizeInBits, cfKeySizeValue);

#ifndef NO_KEYCHAIN
        CFDictionaryAddValue(attributes, kSecUseKeychain, tempKeychain);
#endif

        status = SecKeyGeneratePair(attributes, pPublicKey, pPrivateKey);

        if (status == noErr)
        {
            status = ExportImportKey(pPublicKey, kSecItemTypePublicKey);
        }

        if (status == noErr)
        {
            status = ExportImportKey(pPrivateKey, kSecItemTypePrivateKey);
        }
    }
    else
    {
        status = errSecAllocate;
    }

    if (attributes != NULL)
        CFRelease(attributes);
    if (cfKeySizeValue != NULL)
        CFRelease(cfKeySizeValue);

    *pOSStatus = status;
    return status == noErr;
}
```

Reply:
@bartonjs  Jeremy, I tried enabling `AppleCryptoNative_RsaGenerateKey`, but I have no option to proceed with encrypt/decrypt operations b/c they use `SecTransformRef` and others from the security framework for which I was not able to find appropriate replacement.

Reply:
Jeremy, is it possible to exclude or replace the method like `SecTransformRef `, going from the security framework, in order to enable RSA encrypt/decrypt operation? Otherwise I will have to revert that "scalpel" change enabling some native part and just add warning suppression in the managed code.

Reply:
It looks like for encryption the iOS function is not available on macOS, and vice versa.  For mobile it's https://developer.apple.com/documentation/security/1617956-seckeyencrypt?language=objc.

Reply:
I came back to the state w/o enabling RSA-related native part